### PR TITLE
make it possible for retry() to strip an invalid session

### PR DESCRIPTION
**Pull request by jgravois**
tl:dr: _super_ edge case fix.

---
in Hub land we started encountering errors when passing a valid (www.arcgis.com) authentication session to _public_ services hosted by unfederated instances of ArcGIS Server.

because of this currently we're sending a whole lot of anonymous requests _after_ users authenticate and using retry() to tack on a session we had the whole time whenever we encounter an `ArcGISAuthError` (see https://github.com/Esri/ember-arcgis-server-services/pull/46 for more info).

it'd be a whole lot nicer to invert the flow and pass through the authentication we have consistently and `catch` the ones that are public, but unfederated.

this PR would allow us to do that, by ensuring that the original `url` and `options` are passed through.

```js
const url = `https://sampleserver6.arcgisonline.com/arcgis/rest/services/WorldTimeZones/MapServer`;
request(url, {
    authentication,
    params: {
      foo: 'bar'
    }
  })
    .catch(err => {
      if (err.name === 'ArcGISAuthError') {
        // instead of supplying auth, strip it 
        err.retry(() => Promise.resolve(undefined))
          .then(response)
      }
    })
```

### CODE REVIEW

File: packages/arcgis-rest-request/src/request.ts [hunk 1]
```diff
@@ -156,9 +160,13 @@ export function request(
...
   return (authentication
-    ? authentication.getToken(url, {
-        fetch: options.fetch
-      })
+    ? authentication.getToken(
+        url,
+        {
+          fetch: options.fetch
+        },
+        options
```
Comment (tomwayson) [inline packages/arcgis-rest-request/src/request.ts]: First, doesn't this need to be `originalRequestOptions: options`?

Also, the third arg is now being passed regardless of whether or not `authentication` is a `UserSession` or any other type of `IAuthenticationManager`, but we haven't added `originalRequestOptions` to `IAuthenticationManager`'s `getToken()` signature. I _think_ that's technically OK, but feels a little wonky.

File: packages/arcgis-rest-auth/src/UserSession.ts [hunk 2]
```diff
@@ -793,7 +802,10 @@ export class UserSession implements IAuthenticationManager {
           if (!new RegExp(response.owningSystemUrl, "i").test(this.portal)) {
             throw new ArcGISAuthError(
               `${url} is not federated with ${this.portal}.`,
-              "NOT_FEDERATED"
+              "NOT_FEDERATED",
+              response,
+              url,
+              originalRequestOptions
```
Comment (tomwayson) [inline packages/arcgis-rest-auth/src/UserSession.ts]: can we just pass `requestOptions` here instead of needing a new argument?

**Comment by jgravois** [comment]
@tomwayson was right. we can just `catch` and release the AuthError from request().

🎣🐠🐟

**Comment by patrickarlt** [comment]
@jgravois I would like to review this too. I should have it done before the end of the day.

**Comment by jgravois** [comment]
sounds good @patrickarlt. take your time.

**Comment by jgravois** [comment]
rebased (and merged). thank you both!
